File: 2023_python/day1/puzzle2.py
import sys
import os
import re
sys.path.append(os.path.join(os.path.dirname(__file__)))
from puzzle1 import Puzzle1  # noqa: E402


class Puzzle2(Puzzle1):
    text_digits = [
        'zero',
        'one',
        'two',
        'three',
        'four',
        'five',
        'six',
        'seven',
        'eight',
        'nine',
    ]
    reversed_text_digits = [
        'orez',
        'eno',
        'owt',
        'eerht',
        'ruof',
        'evif',
        'xis',
        'neves',
        'thgie',
        'enin',
    ]

    def __init__(self):
        super().__init__()

    def text_to_digit(self, text):
        try:
            return str(self.text_digits.index(text))
        except Exception:
            return '0'

    def reversed_text_to_digit(self, text):
        try:
            return str(self.reversed_text_digits.index(text))
        except Exception:
            return '0'

    def get_first_digit(self, line):
        match = re.search('([0-9]{1})|(zero|one|two|three|four|five|six|seven|eight|nine)', line)
        # print('group1', match.group(1))
        # print('group2', match.group(2))
        if match:
            if match.group(1):
                return match.group(1)
            else:
                return self.text_to_digit(match.group(2))
        return '0'

    def get_last_digit(self, line):
        reversed_line = line[::-1]
        match = re.search('([0-9]{1})|(orez|eno|owt|eerht|ruof|evif|xis|neves|thgie|enin)', reversed_line)
        if match:
            if match.group(1):
                return match.group(1)
            else:
                return self.reversed_text_to_digit(match.group(2))
        return '0'

File: 2023_node/src/day01/puzzle2.ts
import { Puzzle1 } from "./puzzle1";

const spelledDigits = [
  'one',
  'two',
  'three',
  'four',
  'five',
  'six',
  'seven',
  'eight',
  'nine',
];

export class Puzzle2 extends Puzzle1 {

  reverseString(str: string): string {
    return str.split('').reverse().join('');
  }

  spelledToDigit(spelled: string): string {
    return ((spelledDigits.indexOf(spelled) || 0) + 1).toString();
  }

  getFirstDigit(line: string): string {
    const matches = line.match(/([0-9]{1}|one|two|three|four|five|six|seven|eight|nine)/g);

    if (matches && matches[0]) {
      // matches[0].length === 1 meaning the match was a digit (single character) instead of a spelled number (3 or more characters)
      return matches[0].length === 1 ? matches[0] : this.spelledToDigit(matches[0]);
    }
    return '0';
  }

  getLastDigit(line: string): string {
    const reversedLine = this.reverseString(line);
    const matches = reversedLine.match(/([0-9]{1}|eno|owt|eerht|ruof|evif|xis|neves|thgie|enin)/g);
    if (matches && matches[0]) {
      return matches[0].length === 1 ? matches[0] : this.spelledToDigit(this.reverseString(matches[0]));
    }
    return '0';
  }

};
